File: findcontour.cpp
#include "opencv2/imgcodecs.hpp"
#include "opencv2/highgui.hpp"
#include "opencv2/imgproc.hpp"
#include <iostream>
using namespace cv;
using namespace std;
RNG rng(12345);
Mat src;
int nbPoint = 100;
float w1 = 0.6442;
float w2 = 0.3558;
float bias = 6.4676;
int flag = 0;
struct bigger
{
    template<class T>
    bool operator()(T const &a, T const &b) const { return a > b; }
};

Mat img2Canny(Mat img){
    float mthreshold = mean(img)[0];
    Mat thresh;
    medianBlur(img, thresh, 5);
    threshold(thresh, thresh, mthreshold, 0, THRESH_TOZERO);
    normalize(thresh, thresh, 0, 255, NORM_MINMAX, CV_8U);
    Canny(thresh, thresh, 30, 60);
    return thresh;
}


float calTemperature(vector<Point> contour){
    Rect bound = boundingRect(contour);
    Mat lena1 = Mat::zeros(Size(src.cols, src.rows),CV_8UC1);
    rectangle(lena1, bound, 255, -1);
    vector<Point> indices;
    findNonZero(lena1, indices);

    vector<float> vectorTemp;
    float sum = 0;
    float maxTemp = -1;
    for (int i = 0; i<indices.size(); i++){
        __int16_t temp = src.at<__int16_t>(indices[i].x, indices[i].y);
        float celcius = float(temp) / 100 - 273.15; 
        sum += celcius;
        vectorTemp.push_back(celcius);
        if (celcius > maxTemp){
            maxTemp = celcius;
        }
    }

    sort(vectorTemp.begin(), vectorTemp.end(), bigger());
    float maxAverage = 0.0;
    float minAverage = 0.0;
    for (auto i = vectorTemp.begin(); i < vectorTemp.begin()+nbPoint; ++i){
        maxAverage += *i;
    }

    for (auto i = vectorTemp.end() - 100; i < vectorTemp.end(); i++)
    {
        minAverage += *i;
    }
    float maxBoundingboxTemp = maxAverage / nbPoint;
    float environmentTemp = minAverage / nbPoint;
    cout << "average temperature of human is " << maxBoundingboxTemp << endl;
    cout << "average temperature of environment is " << environmentTemp << endl;
    cout << "bouding box size: " << indices.size() << endl;
    float averageTemp = sum / indices.size();
    cout << "Average temp of contour "  << averageTemp << endl;
    cout << "Max temp " << maxTemp << endl;
    cout << "Min temp " << vectorTemp.back() << endl;

    if (flag==1){
        float predictTemp = w1 * maxBoundingboxTemp + w2 * environmentTemp + bias;
        cout << "predict temperature of this image is: " << predictTemp << endl;
    }

    cout << endl;
}

void drawMaxContour(Mat img, Mat dst){
    vector<vector<Point> > contours;
    vector<Vec4i> hierarchy;

    int maxIndex = -1;
    float maxArea = -1.0;    
    findContours(img, contours, RETR_TREE, CHAIN_APPROX_NONE);
    for (int i=0; i<contours.size(); i++){
        vector<Point>  hull;
        convexHull(contours[i], hull);
        float area = contourArea(hull);
        if (area > maxArea) {
            maxIndex = i;
            maxArea = area;
        }
    }

    Scalar color = Scalar(0, 255, 0);
    drawContours(dst, contours, maxIndex, color, -1);
    calTemperature(contours[maxIndex]);
    Mat mresize;
    resize(dst, mresize, Size(640, 480), INTER_LINEAR);
    if (flag==1) {
        imshow("dst", mresize);
        waitKey(0);
    }
    // imshow("lena", lena1);
}


int main( int argc, char** argv )
{   
    const String keys = "{@input | ./IR/1595432616.161967.png | input image}"
                        "{image  | 1                          | showimage}";
 
    CommandLineParser parser( argc, argv,  keys);
    // int mode[10] = {IMREAD_COLOR, IMREAD_UNCHANGED, IMREAD_GRAYSCALE, IMREAD_ANYDEPTH, };
    flag = parser.get<int>("image");
    src = imread( samples::findFile( parser.get<String>( "@input" ) ) , IMREAD_ANYDEPTH);

    if( src.empty() )
    {
      cout << "Could not open or find the image!\n" << endl;
      cout << "Usage: " << argv[0] << " <Input image>" << endl;
      return -1;
    }

    Mat norm, drawing;
    normalize(src, norm, 0, 255, NORM_MINMAX, CV_8U); // for visualizing 
    // (drawing, CV_8UC3);
    cvtColor(norm, drawing, COLOR_GRAY2BGR);
    Mat thresh = img2Canny(src);
    drawMaxContour(thresh, drawing);
    // imshow("canny", norm);
    
    return 0;
}

File: findcontour.py
import cv2
import numpy as np
import random as rng
import sys
import matplotlib.pyplot as plt

def read_image(file):
    img = cv2.imread(file, cv2.IMREAD_ANYDEPTH)
    thresh = cv2.medianBlur(img, 5)
    thresh_hold = np.mean(thresh)
    ret, thresh = cv2.threshold(src=thresh, thresh=thresh_hold, maxval=0, type=cv2.THRESH_TOZERO)
    normed = cv2.normalize(thresh, None, 0, 255, cv2.NORM_MINMAX, cv2.CV_8U)
    out_canny = cv2.Canny(normed, 30, 60)
    return out_canny, img


def find_contour(image):
    contours, hierarchy = cv2.findContours(image=image, mode=cv2.RETR_TREE, method=cv2.CHAIN_APPROX_NONE)
    max_val = -1
    max_are = -1
    # print(len(contours))
    for i in range(len(contours)):
        area = cv2.contourArea(cv2.convexHull(contours[i]))
        # print(area)
        if area > max_are:
            max_val = i
            max_are = area
    return contours, max_val, hierarchy

img_path = sys.argv[1]
print(img_path)

img = cv2.imread(img_path, cv2.IMREAD_ANYDEPTH)
if img is None:
    print('not found img')
    exit()

canny, ori = read_image(img_path)
contours, max_val, hierarchy = find_contour(canny)

normed = cv2.normalize(ori, None, 0, 255, 32, cv2.CV_8U)
normed_rgb = cv2.cvtColor(normed, cv2.COLOR_GRAY2BGR)

cv2.drawContours(normed_rgb, contours, max_val, (0, 255, 0), 2, cv2.LINE_8, hierarchy, 0)
fig = plt.figure(figsize=(12, 8))
fig.add_subplot(1,2,1)
plt.imshow(ori, cmap="hot");
plt.title("origin image")
fig.add_subplot(1,2,2)
plt.imshow(normed_rgb)
plt.title("find contour image")
plt.show()

# cv2.imshow("contour", cv2.resize(normed_rgb, dsize=(640, 480), interpolation=cv2.INTER_LINEAR))
# cv2.waitKey(0)
# cv2.destroyAllWindows()
